altair/utils/transformed_data.py

```diff
@@ -0,0 +1,545 @@
+) -> List[str]:
+    """Name unnamed charts and subcharts
+
+    Name unnamed charts and subcharts so that we can look them up later in
+    the compiled Vega spec.
+
+    Note: This function mutates the input chart by applying names to
+    unnamed charts.
+
+    Parameters
+    ----------
+    chart : Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, or ConcatChart
+        Altair chart to apply names to
+    i : int (default 0)
+        Starting chart index
+    exclude : iterable of str
+        Set of the names of charts to exclude
+
+    Returns
+    -------
+    list of str
+        List of the names of the charts and subcharts
+    """
+    exclude = set(exclude) if exclude is not None else set()
+    if isinstance(chart, (Chart, FacetChart)):
+        # Perform shallow copy of chart so that we can change
+        # the name
+        if chart.name not in exclude:
+            if chart.name in (None, Undefined):
+                name = make_magic_chart_name(i)
+                chart.name = name
+            return [chart.name]
+        else:
+            return []
+    else:
+        if isinstance(chart, LayerChart):
+            subcharts = chart.layer
+        elif isinstance(chart, HConcatChart):
+            subcharts = chart.hconcat
+        elif isinstance(chart, VConcatChart):
+            subcharts = chart.vconcat
+        elif isinstance(chart, ConcatChart):
+            subcharts = chart.concat
+        else:
+            raise ValueError(
+                "transformed_data accepts an instance of "
+                "Chart, FacetChart, LayerChart, HConcatChart, VConcatChart, or ConcatChart\n"
+                f"Received value of type: {type(chart)}"
+            )
+
+        chart_names: List[str] = []
+        for subchart in subcharts:
+            for name in name_chart(subchart, i=i + len(chart_names), exclude=exclude):
+                chart_names.append(name)
+        return chart_names
+
+
+def get_group_mark_for_scope(vega_spec: dict, scope: Scope) -> Optional[dict]:
+    """Get the group mark at a particular scope
+
+    Parameters
+    ----------
+    vega_spec : dict
+        Top-level Vega specification dictionary
+    scope : tuple of int
+        Scope tuple. If empty, the original Vega specification is returned.
+        Otherwise, the nested group mark at the scope specified is returned.
+
+    Returns
+    -------
+    dict or None
+        Top-level Vega spec (if scope is empty)
+        or group mark (if scope is non-empty)
+        or None (if group mark at scope does not exist)
+
+    Examples
+    --------
+    >>> spec = {
+    ...     "marks": [
+    ...         {
+    ...             "type": "group",
+    ...             "marks": [{"type": "symbol"}]
+    ...         },
+    ...         {
+    ...             "type": "group",
+    ...             "marks": [{"type": "rect"}]}
+    ...     ]
+    ... }
+    >>> get_group_mark_for_scope(spec, (1,))
+    {'type': 'group', 'marks': [{'type': 'rect'}]}
```

Comment:
Does `hatch run test` automatically run doctests?

Reply:
`hatch run test` runs the following: 
```cmd
black --diff --color --check .
ruff check .
mypy altair tests
python -m pytest --pyargs --doctest-modules tests
```
See https://github.com/altair-viz/altair/blob/master/pyproject.toml#L104-L107